storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,16 +3579,326 @@ dberr_t fsp_sys_tablespace_validate()
 }
 #endif /* UNIV_DEBUG */
 
-void fsp_system_tablespace_truncate()
+/** Store the inode information which basically stores
+the page and offset */
+struct inode_info : private std::unordered_set<uint64_t>
+{
+public:
+  __attribute__((warn_unused_result))
+  /** Register an inode */
+  bool insert_seg(const byte *inode)
+  {
+    uint32_t page= mach_read_from_4(inode + 4);
+    uint32_t offset= mach_read_from_2(inode + 8);
+    return page >= fil_system.sys_space->free_limit ||
+      offset < FIL_PAGE_DATA || offset >= srv_page_size - FIL_PAGE_DATA_END ||
+      emplace(uint64_t{page} << 16 | offset).second;
+  }
+
+  __attribute__((warn_unused_result))
+  bool find(uint32_t page, uint16_t offset) const
+  { return std::unordered_set<uint64_t>::find(uint64_t{page} << 16 | offset) != end(); }
+};
+
+/** Determine the inodes used by tables in the system tablespace.
+@param mtr    mini-transaction
+@param inodes index nodes found
+@return error code */
+static dberr_t fsp_table_inodes(mtr_t *mtr, inode_info *inodes)
+{
+  btr_pcur_t pcur;
+  ulint len;
+  const auto savepoint= mtr->get_savepoint();
+  dberr_t err= DB_SUCCESS;
+  dict_sys.freeze(SRW_LOCK_CALL);
+  std::unordered_set<uint32_t> root_pages;
+  for (const rec_t *rec= dict_startscan_system(
+         &pcur, mtr, dict_sys.sys_indexes);
+       rec; rec= dict_getnext_system_low(&pcur, mtr))
+  {
+    const byte *field= rec_get_nth_field_old(
+      rec, DICT_FLD__SYS_INDEXES__SPACE, &len);
+    if (len != 4)
+    {
+      err= DB_CORRUPTION;
+      break;
+    }
+    uint32_t space= mach_read_from_4(field);
+    if (space > 0) continue;
+
+    field= rec_get_nth_field_old(
+      rec, DICT_FLD__SYS_INDEXES__PAGE_NO, &len);
+    if (len != 4)
+    {
+      err= DB_CORRUPTION;
+      break;
+    }
+    uint32_t root_page_no= mach_read_from_4(field);
+    if (root_page_no != FIL_NULL)
+      root_pages.insert(root_page_no);
```

Comment:
We are ignoring duplicate `root_pages` here. Can we avoid introducing a local variable `root_pages` and update `inodes` directly? It would seem to require a further function to be invoked here:
```c++
  uint32_t root= mach_reaad_from_4(field);
  if (root != FIL_NULL && (err= fsp_table_inodes_root(mtr, root) != DB_SUCCESS)
    break;
```